CIPS/cip-2.md

```diff
@@ -0,0 +1,50 @@
+# Distribution module 2.0: Farming the Stars
+
+## Motivation
+During the original launch of the Cosmos SDK, the distribution module was  one of the key breakthroughs behind Cosmos proof of stake. It enables the block by block distribution of staking rewards and fees to millions of delegators in a scalable manner.
+
+Buried within the distribution module is something much more powerful, a generic reward module that can enable Cosmos SDK application developers to enable yield farming, rewards for Interchain staking and more. The goal of this endeavor is to surface this potential.
+
+# Use Cases
+### Staking Derivatives
+
+Staking derivatives requires a few components to be established. 
+
+1. Staking derivatives reworking the mechanics of the staking module and distribution module to respond to possession of the appropriate denom for both withdrawing staked tokens and collecting rewards
+2. Some staking derivatives designs require reward behavior to be triggered at the epoch boundary. The distribution module should be adapted to integrate with epoch base staking.
```

Comment:
without going too much into a technical discussion, I think we should not require that the state update is happening at each epoch. If eapoch=block, then it can be super expensive: we will need to query all staking users and update them.

Instead, we can think about an optimistic approach:
* update a user balance only during related user transactions (so if user X is doing something which impacts his stake, then we will update the balance)
* expose to a user a `ping` method which will allow him to trigger an update on demand.

One limitation of this design is that it's harder to combine it with a logic based on a total staked asset. For example: let's say we are farming 100 tokens a day, and we distribute it to all users proportionally to their stake. The stake token = farming token and let's say that we require automatic re-staking of the farmed tokens. Now we will have difficulty to compute a total stake, because we don't update all users balance with compounding "revenues" in a real time.